File: src/bind/gp.cpp
#include "bind.h"

#include "../gp/ax1.h"
#include "../gp/ax2.h"
#include "../gp/glm.h"

static glm::vec3 xyzAdded(const glm::vec3 &xyz1, const glm::vec3 &xyz2) {
  return xyz1 + xyz2;
}

static glm::vec3 xyzSubtracted(const glm::vec3 &xyz1, const glm::vec3 &xyz2) {
  return xyz1 - xyz2;
}

static glm::vec3 xyzMultiplied(const glm::vec3 &xyz, glm::f32 scale) {
  return xyz * scale;
}

static glm::vec3 xyzDivided(const glm::vec3 &xyz, glm::f32 scale) {
  return xyz / scale;
}

static glm::vec3 xyzNegated(const glm::vec3 &xyz) {
  return -xyz;
}

void lviz::bind::BindGP(lua_State *L) {
  luabridge::getGlobalNamespace(L)
      .beginNamespace("lviz")
      .beginNamespace("gp")
      .beginClass<gp::Pnt>("Pnt")
      .addConstructor<void(), void(const glm::vec3 &),
                      void(float, float, float)>()
      .addProperty("x", &gp::Pnt::x)
      .addProperty("y", &gp::Pnt::y)
      .addProperty("z", &gp::Pnt::z)
      .addFunction("IsUnset", &gp::IsUnsetXYZ)
      .addFunction("Distance", &gp::PntDistance)
      .addFunction("Distance2", &gp::PntDistance2)
      .addFunction("Added", &xyzAdded)
      .addFunction("Subtracted", &xyzSubtracted)
      .addFunction("Multiplied", &xyzMultiplied)
      .addFunction("Divided", &xyzDivided)
      .addFunction("Negated", &xyzNegated)
      .endClass()
      .beginClass<gp::Vec>("Vec")
      .addConstructor<void(), void(const glm::vec3 &),
                      void(float, float, float)>()
      .addProperty("x", &gp::Vec::x)
      .addProperty("y", &gp::Vec::y)
      .addProperty("z", &gp::Vec::z)
      .addFunction("IsUnset", &gp::IsUnsetXYZ)
      .addFunction("Magnitude", &gp::VecMagnitude)
      .addFunction("Added", &xyzAdded)
      .addFunction("Subtracted", &xyzSubtracted)
      .addFunction("Multiplied", &xyzMultiplied)
      .addFunction("Divided", &xyzDivided)
      .addFunction("Negated", &xyzNegated)
      .addFunction("Dot", &gp::VecDot)
      .addFunction("Crossed", &gp::VecCrossed)
      .addFunction("Angle", &gp::VecAngle)
      .addFunction("AngleWithRef", &gp::VecAngleWithRef)
      .endClass()
      .beginClass<gp::Ax1>("Ax1")
      .addConstructor<void(), void(const gp::Pnt &, const gp::Vec &)>()
      .addFunction("Location", &gp::Ax1::Location)
      .addFunction("Direction", &gp::Ax1::Direction)
      .endClass()
      .beginClass<gp::Ax2>("Ax2")
      .addConstructor<void(), void(const gp::Pnt &, const gp::Vec &),
                      void(const gp::Pnt &, const gp::Vec &, const gp::Vec &)>()
      .addFunction("Axis", &gp::Ax2::Axis)
      .addFunction("Location", &gp::Ax2::Location)
      .addFunction("Direction", &gp::Ax2::Direction)
      .addFunction("XDirection", &gp::Ax2::XDirection)
      .addFunction("YDirection", &gp::Ax2::YDirection)
      .addFunction("ToWorld", &gp::Ax2::ToWorld)
      .addFunction("ToLocal", &gp::Ax2::ToLocal)
      .endClass()
      .endNamespace()
      .endNamespace();
}

File: runtime/lua/_meta/gp.lua
---@meta _
error("Cannot require a meta file")

lviz.gp = {}

---@class lviz.gp.Pnt
---@field x number
---@field y number
---@field z number
---@overload fun():lviz.gp.Pnt
---@overload fun(other: lviz.gp.Pnt|lviz.gp.Vec):lviz.gp.Pnt
---@overload fun(x: number, y: number, z: number):lviz.gp.Pnt
lviz.gp.Pnt = {}

---
---@return boolean
function lviz.gp.Pnt:IsUnset() end

---
---@param other lviz.gp.Pnt
---@return number
function lviz.gp.Pnt:Distance(other) end

---
---@param other lviz.gp.Pnt
---@return number
function lviz.gp.Pnt:Distance2(other) end

---
---@param other lviz.gp.Pnt|lviz.gp.Vec
---@return lviz.gp.Pnt
function lviz.gp.Pnt:Added(other) end

---
---@overload fun(self: lviz.gp.Pnt, other: lviz.gp.Pnt):lviz.gp.Vec
---@overload fun(self: lviz.gp.Pnt, other: lviz.gp.Vec):lviz.gp.Pnt
function lviz.gp.Pnt:Subtracted(...) end

---
---@param scale number
---@return lviz.gp.Pnt
function lviz.gp.Pnt:Multiplied(scale) end

---
---@param scale number
---@return lviz.gp.Pnt
function lviz.gp.Pnt:Divided(scale) end

---
---@return lviz.gp.Pnt
function lviz.gp.Pnt:Negated() end

---@class lviz.gp.Vec
---@field x number
---@field y number
---@field z number
---@overload fun():lviz.gp.Vec
---@overload fun(other: lviz.gp.Pnt|lviz.gp.Vec):lviz.gp.Vec
---@overload fun(x: number, y: number, z: number):lviz.gp.Vec
lviz.gp.Vec = {}

---
---@return boolean
function lviz.gp.Vec:IsUnset() end

---
---@return number
function lviz.gp.Vec:Magnitude() end

---
---@param other lviz.gp.Vec
---@return lviz.gp.Vec
function lviz.gp.Vec:Added(other) end

---
---@param other lviz.gp.Vec
---@return lviz.gp.Vec
function lviz.gp.Vec:Subtracted(other) end

---
---@param scale number
---@return lviz.gp.Vec
function lviz.gp.Vec:Multiplied(scale) end

---
---@param scale number
---@return lviz.gp.Vec
function lviz.gp.Vec:Divided(scale) end

---
---@return lviz.gp.Vec
function lviz.gp.Vec:Negated() end

---
---@param other lviz.gp.Vec
---@return number
function lviz.gp.Vec:Dot(other) end

---
---@param other lviz.gp.Vec
---@return lviz.gp.Vec
function lviz.gp.Vec:Crossed(other) end

---
---@param other lviz.gp.Vec
---@return number
function lviz.gp.Vec:Angle(other) end

---
---@param other lviz.gp.Vec
---@param ref lviz.gp.Vec
---@return number
function lviz.gp.Vec:AngleWithRef(other, ref) end

---@class lviz.gp.Ax1
---@overload fun():lviz.gp.Ax1
---@overload fun(location: lviz.gp.Pnt, direction: lviz.gp.Vec):lviz.gp.Ax1
lviz.gp.Ax1 = {}

---
---@return lviz.gp.Pnt
function lviz.gp.Ax1:Location() end

---
---@return lviz.gp.Vec
function lviz.gp.Ax1:Direction() end

---@class lviz.gp.Ax2
---@overload fun():lviz.gp.Ax2
---@overload fun(location: lviz.gp.Pnt, direction: lviz.gp.Vec):lviz.gp.Ax2
---@overload fun(location: lviz.gp.Pnt, x_direction: lviz.gp.Vec, y_direction: lviz.gp.Vec):lviz.gp.Ax2
lviz.gp.Ax2 = {}

---
---@return lviz.gp.Ax1
function lviz.gp.Ax2:Axis() end

---
---@return lviz.gp.Pnt
function lviz.gp.Ax2:Location() end

---
---@return lviz.gp.Vec
function lviz.gp.Ax2:Direction() end

---
---@return lviz.gp.Vec
function lviz.gp.Ax2:XDirection() end

---
---@return lviz.gp.Vec
function lviz.gp.Ax2:YDirection() end

---
---@param local_coord lviz.gp.Pnt
---@return lviz.gp.Pnt
function lviz.gp.Ax2:ToWorld(local_coord) end

---
---@param world_coord lviz.gp.Pnt
---@return lviz.gp.Pnt
function lviz.gp.Ax2:ToLocal(world_coord) end
